tests/src/gui/testqgsdoublevalidator.cpp

```diff
@@ -0,0 +1,185 @@
+/***************************************************************************
+    testqgsdoublespinbox.cpp
+     --------------------------------------
+    Date                 : June 2020
+    Copyright            : (C) 2020 Sebastien Peillet
+    Email                : sebastien dot peillet at oslandia dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include "qgstest.h"
+
+#include "qgsdoublevalidator.h"
+#include <QLineEdit>
+
+class TestQgsDoubleValidator: public QObject
+{
+    Q_OBJECT
+  private slots:
+    void initTestCase(); // will be called before the first testfunction is executed.
+    void cleanupTestCase(); // will be called after the last testfunction was executed.
+    void init(); // will be called before each testfunction is executed.
+    void cleanup(); // will be called after every testfunction.
+
+    void validate();
+    void validate_data();
+    void toDouble_data();
+    void toDouble();
+
+  private:
+
+};
+
+void TestQgsDoubleValidator::initTestCase()
+{
+
+}
+
+void TestQgsDoubleValidator::cleanupTestCase()
+{
+}
+
+void TestQgsDoubleValidator::init()
+{
+}
+
+void TestQgsDoubleValidator::cleanup()
+{
+}
+
+void TestQgsDoubleValidator::validate_data()
+{
+  QTest::addColumn<QString>( "actualState" );
+  QTest::addColumn<int>( "expState" );
+
+  QTest::newRow( "C decimal" ) << QString( "4cd6" ) << int( QValidator::Acceptable );
+  QTest::newRow( "locale decimal" ) << QString( "4ld6" ) << int( QValidator::Acceptable );
+  QTest::newRow( "locale decimal" ) << QString( "4444ld6" ) << int( QValidator::Acceptable );
+
+  // QgsDoubleValidator doesn't expect group separator but it tolerates it,
```

Comment:
why is that considered non acceptble? 

If the user enters the numbr `1234.56789` in it_IT locale as `1.234,567899` this would be correctly converted by `QLocale::toDouble()` without any required editing by the user, so I think it should be valid.

Reply:
Actually it's not "acceptable" because of the regular expression constraint,  but the value will be converted. The actual regexp doesn't include group separator. The user can't enter a number using group separator, except when he pastes the whole number.

Intermediate validation is based on the `QLocale::toDouble()` result (so Intermediate result can be converted) and to get Acceptable result, it needs to fit regexp condition and range limit.

I choose to put aside group separator to ensure that the entered value will be converted no matter what. For example, I wish to enter the number 1234567.89 in it_IT locale as 1.234.567,89 from the left to the right : 

|String value | Result | comment|
|-----------------|----------|-------------|
|1 | 1 |it_IT locale |
|1. | 1 |C locale |
|1.2 | 1.2 |C locale |
|1.23 | 1.23 |C locale |
|1.234 | 1234 |it_IT locale |
|1.234. & 1.234.5 | 0 | can't be converted as it excepted a number like x.xxx.xxx in it_IT locale or x.xxxxxxxx in C locale. At this point the entered value it can't be converted and does not fit regexp. To avoid that user enters wrong value the validator does not allow to add the second '.' without the 3 digits at the same time as it will be possible to convert it.
|1.234.56 | 0 | can't be converted and does not fit regexp but allow to add an extra digit as it will be possible to convert the string.|
1.234.567 | 1234567 |it_IT locale|
1.234.567,8 |1234567.8 | it_IT locale|
1.234.567,89 |1234567.89 | it_IT locale|

So in conclusion @elpaso, you can paste value with group separator and it will be converted in double value if you validate the form. Maybe we should be more flexible on the group separator but it could lead to wrong value and on my opinion QValidator exists to avoid these mistakes

Reply:
That makes sense, thank you for clarifying.